README.md
# CODATA fundamental physical constants

**Language-specific CODATA fundamental physical constants files, generated by a Python script from the NIST website.**

The `nist.py` Python script downloads the CODATA recommended values (latest by default), available as an ASCII file on the NIST website:

https://physics.nist.gov/cuu/Constants/Table/allascii.txt

and generates a language-specific file containing these values and available for computations. The 2018 CODATA set contains 354 constants. The next set should be published in 2022. The code currently outputs constants for the following languages:
1. `Fortran` (including extensive testing) `CODATA_constants.f90`
2. `C/C++` header file `CODATA_constants.h`
3. `Python` file `CODATA_constants.py`
4. `Matlab` file `CODATA_constants.m` with an autosave as `CODATA_constants.mat`
5. `Mathematica` file `CODATA_constants.copy_to_nb` to be directly copied into Mathematica notebook.
## Generating a new file

### Default options

From the project root directory, type:

```bash
$ ./nist.py
Generating the fortran/CODATA_constants.f90 file using the recommended 2018 values:
https://physics.nist.gov/cuu/Constants/Table/allascii.txt
354 constants written
34423 bytes
```

### Options of the Python script

```bash
$ ./nist.py --help
usage: nist.py [-h] [-y {2010,2014,2018}] [-d] [-v] [-l {fortran,c,cpp,python,matlab,mathematica}]

Downloads CODATA fundamental physical constants from NIST website and generates a language-specific file (including a Fortran module).

optional arguments:
  -h, --help           show this help message and exit
  -y {2010,2014,2018}  CODATA values: 2010, 2014 or 2018
  -d                   Delete calculated values (...)
  -v                   Version
  -l                   Output language name: fortran, c, cpp, python, matlab, mathematica

Sources : <https://github.com/chemical-accuracy/fundamental_constants>
```

By default latest values are downloaded. But you can force the year with the `-y` option.

By default, calculated values are kept. But you can delete them with the `-d` option if you prefer to calculate them in your program, in order to have more decimals.

### Why is there an option for all the different years?

Most of the time you will use the latest values. But if you need to reproduce exactly the results obtained several years ago, you may need to use the corresponding CODATA set. It can be also useful if your project contains tests that expect a very precise numerical result.

## Fortran
For Fortran `module CODATA_constants` in the file `fortran/CODATA_constants.f90` is created. It also generates a Fortran program `test/test.f90` which uses that module to generate a text file reproducing the two first columns, with (nearly) the same formatting as in the CODATA NIST file.

### Ready to use validated Fortran modules

In the `fortran/` directory, you will find for each recent CODATA sets:

- the corresponding validated `CODATA_????_constants.f90` module,
- the `fortran_generated_????.txt` file generated using this module. A graphical diff tool was used to verify that the numbers printed in that file are identical to numbers in the NIST file, which validates the Fortran module,
- the `CODATA_latest_constants.f90` module, if you are only interested by the latest set.

If you use the Fortran Package Manager fpm, you can add in the "dependencies" section of your `fpm.toml` file:

```toml
fundamental_constants = { git="https://github.com/chemical-accuracy/fundamental_constants.git" }
```

or you can simply copy/paste the `.f90` files you need in your project.


### Using the Fortran modules

The constant names in the modules are quite long:

```Fortran
module CODATA_2018_constants
  use, intrinsic :: iso_fortran_env, only: wp=>real64
  implicit none

  real(wp), parameter :: alpha_particle_electron_mass_ratio = 7294.29954142_wp ! 0.000 000 24  
  real(wp), parameter :: alpha_particle_mass = 6.6446573357e-27_wp ! 0.000 000 0020 e-27  kg
  real(wp), parameter :: alpha_particle_mass_energy_equivalent = 5.9719201914e-10_wp ! 0.000 000 0018 e-10  J
...
```

You should rename them upon import like this:

```fortran
use CODATA_latest_constants, only: c=>speed_of_light_in_vacuum
```

You can create your own short module with a `use CODATA_latest_constants` statement and put into it your renamed favourite constants, accordingly to what you generally need in your scientific field.

Note also that the names of some constants may vary each time CODATA releases new recommended values.



## Validating the values in the Fortran module

If you have the fpm Fortran Package Manager, build and launch the test:

```bash
$ fpm test
[...]
 Do you know this constant with so many zeros?
0100000110110001110111100111100001001010000000000000000000000000
 Oh sorry, you\'re that kind of biped with 10 digits?
   299792458.00000000     
 Some other constants:
   6.0221407599999999E+023
   1.3806490000000001E-023
   1.6021766339999999E-019
 4.500 3878060E-10        
 2808.92113298            
 1234.123 456 78          
 Writing the fortran_generated_????.txt test file
```

A graphical diff tool, like `meld`, can be used to compare the `fortran_generated_????.txt` file with the NIST file, in order to validate the Fortran module. Most constants are written exactly as in the NIST file. A few constants are written differently because the NIST file don't use the classical scientific format: for example, the molar mass of carbon-12 is written 1.199 999 999 58 e-2 in our file but 11.99 999 999 58 e-3 in the NIST file. But it is quick to verify that those values are correct. Some of the values in the NIST file can be calculated from other constants and are terminated by "..." in the NIST file: note that their values in the Fortran module are not calculated, but have the same number of digits as in the NIST file.

You can use alternatively the `build.sh` script to build the `build/runTests` executable.


## References
* Project forked from https://github.com/vmagnin/fundamental_constants
* Origin of this project: https://fortran-lang.discourse.group/t/physical-constants/822
* About CODATA:
    * https://physics.nist.gov/cuu/Constants/bibliography.html
    * https://en.wikipedia.org/wiki/Committee_on_Data_for_Science_and_Technology
* NIST website: https://physics.nist.gov/cuu/Constants/
* If you need a more complete project around constants, you may be interested by:
    * https://gitlab.com/apthorpe/stdlib-codata
    * https://github.com/MolSSI/QCElemental
## Future plans:
1. Periodic table.
2. Better names for constants.
3. Derived units.

CHANGELOG.md
# Changelog
All notable changes to the project will be documented in this file.
The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).

## [fundamental constants v1.11] 2022-07-02

## Added
- -l argument to the main python script (nist.py) allowing user to choose the output language
- Ability to generate a C/C++ header file analogous to the Fortran's MODULE file
- Ability to generate a .py, .m and .copy_to_nb files for Python, Matlab and Mathematica calculations respectively
- Single script `generate.sh` that generates all the versions for all languages

## Changed
- Code now contains match statements, requiring Python 3.10+
- Move fortran-specific features from `.` and `src` to `fortran` directory
- Update the lint to NIST ASCII table (it changed for the older tables)
- Print out the year in the file name if it's not the newest constants
## [fundamental constants v1.1] 2021-04-03

## Added
- Added CODATA_latest_constants.f90 simply using the latest available module (2018 at the moment).

## Changed
- Renamed the Fortran CODATA modules and files and moved them in `src` directory.
- README.md updated.


## [fundamental constants v1.0] 2021-03-23

## Added
- A `validated_modules` directory with 2010, 2014 and 2018 Fortran modules, and the text files generated by `fpm test`.

## Changed
- README.md updated.

## Fixed
- test/functions.f90: fixed a bug concerning six CODATA constants with non classical scientific format.
- nist.py: fixed a bug concerning the 2010 file (lines are terminated by `\r\n`).


## [fundamental constants v0.3] 2021-03-22

### Added
- test/functions.f90:
    - a subroutine `insert_space(string, position)` to insert a space in the string at position (in the range [1, 25]).
    - a subroutine `remove_character(string, position)`.
    - a subroutine `format_digits(string)` to insert spaces between groups of digits.

### Changed
- test/functions.f90:
    - improved `write_CODATA()`. The two columns are nearly identical to the CODATA format.

    
## [fundamental constants v0.2] 2021-03-20

### Added
- nist.py: a function `get_number_parts(xstring)` that returns a tuple with the digits at the left and right of the decimal point, and the value of the exponent.
- test/test.f90: automatically written by nist.py. Prints the name of the constant and its value.
- test/functions.f90: `write_CODATA()` tries to output the name and the value of a constant in the same format as in the CODATA listing.


## [fundamental constants v0.1] 2021-03-13

### Added
- Initial commit.


nist.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
'''
Downloads CODATA fundamental physical constants from NIST website
and generates a language-specific (fortran, C/C++) file. For Fortran, 
module CODATA_constants.f90 and also a test.f90 file are generated.
The latter can be used to validate the process.
For C/C++ a header file CODATA_constants.h is generated
Original code by Vincent MAGNIN, 2021-03-11
MIT license
https://github.com/vmagnin/fundamental_constants
Last modifications: 2022-07-02
'''

import argparse
import os
import datetime
from urllib.request import urlopen

def get_number_parts(xstring):
    '''
    Receives a floating point passed as a string, and returns
    a tuple with the digits at the left and right of the
    decimal point, and the value of the exponent.
    '''
    # Is there an exponent?
    exponent_pos = xstring.find("e")
    if exponent_pos != -1:
        expo = int(xstring[exponent_pos+1:])
        # We remove the exponent
        # (right bound excluded in Python):
        xstring = xstring[0:exponent_pos]
    else:
        expo = 0

    # Number of digits at the left of the decimal point:
    point_pos = xstring.find(".")
    if point_pos == -1:
        left = len(xstring)
    else:
        left = point_pos
    # Must be corrected if a sign is present:
    if xstring[0] in ('-', '+'):
        left = left - 1

    # We keep only the decimals and count them:
    if point_pos != -1:
        xstring = xstring[point_pos+1:]
        right = len(xstring)
    else:
        right = 0

    return [left, right, expo]


#**************
# Main program
#**************

# Command line arguments:
PARSARG = argparse.ArgumentParser(description="""Downloads CODATA recommended 
                                  values of physical constants from NIST 
                                  website and generates a Fortran module.""",
                                  epilog="Sources : <https://github.com/vmagnin/fundamental_constants>")
PARSARG.add_argument("-y", action="store", nargs=1, type=int,
                     choices=[2010, 2014, 2018], default=[2018],
                     help="CODATA values: 2010, 2014 or 2018")
PARSARG.add_argument("-d", action="store_true",
                     help="Delete calculated values (...)")
PARSARG.add_argument("-v", action="version",
                     version="%(prog)s v1.11 MIT license", help="Version")
PARSARG.add_argument("-l", nargs=1, type=str, 
                    choices=['fortran', 'Fortran','FORTRAN','fort','FORT','Fort', # fortran
                    'c','cpp','C','Cpp','CPP','c++','C++', # C and C++
                    'Python', 'python', 'PYTHON', 'py', 'PY', # Python
                    'matlab','Matlab','MATLAB','mat','MAT', # Matlab
                    'mathematica', 'Mathematica' # mathematica
                    ], default=['fortran'],help="Output language name: fortran, c, cpp, python, matlab, mathematica")
ARGS = PARSARG.parse_args()

# Those previous values are not available as an ASCII file:
# 2006, 2002, 1998, 1986, 1973, 1969.
# Next version should be released in 2022.
newest = False
match ARGS.y[0]:
    case 2018:
        newest = True
        url = "https://physics.nist.gov/cuu/Constants/Table/allascii.txt"
    case 2014:
        url = "https://physics.nist.gov/cuu/Constants/ArchiveASCII/allascii_2014.txt"
    case 2010:
        url = "https://physics.nist.gov/cuu/Constants/ArchiveASCII/allascii_2010.txt"

# Downloading the file from NIST website:
with urlopen(url) as nist_file:
    file_content = nist_file.read().decode('utf-8')
    lines_list = file_content.splitlines(True)

MODULE_NAME = "CODATA_" + (str(ARGS.y[0]) + '_')*(not newest) + "constants"
TABS = "  "
fortran=False
folder_out = 'other_languages/'
const_suffix = ' '
const_prefix = final_line = trail_line_comment = ""
match ARGS.l[0].lower():
    case 'fortran':
        fortran=True
        folder_out = 'fortran/'
        extention = ".f90"
        line_comment = '!'
        const_prefix = "  real(wp), parameter :: "
        const_suffix = "_wp "
        final_line = "end module "+MODULE_NAME+"\n"
    case 'c' | 'cpp' | 'c++':
        extention = '.h'
        line_comment = '//'
        const_prefix = "const double "
        const_suffix = ";"
    case 'python' | 'py':
        extention = '.py'
        line_comment = '#'
        final_line = "# Note: there is a MolSSI module for python https://github.com/MolSSI/QCElemental\n" + "# Also physical constants are available from 'scipy.constants'"
    case 'matlab' | 'mat':
        extention = '.m'
        line_comment = '%'
        final_line = "\n% save to a binary MAT file '" + f"{MODULE_NAME}"+".mat', which can later be loaded by your MATLAB script.\n"+f"save {MODULE_NAME}"
    case "mathematica":
        extention = '.copy_to_nb' # this one is not an easy text file
        line_comment = '(*'
        trail_line_comment = ' *)'
        final_line = "(* Copy the contents of this file to a cell of your notebook.nb *)"
    case _: raise Exception("This language is not recognized or has not been implemented yet.")
# Creating the constant-containing file:
out_file_name = MODULE_NAME + extention
out_file = open(folder_out+out_file_name, "w")
print("Generating the "+ folder_out +out_file_name+" file using the recommended "
      + str(ARGS.y[0]) + " values:")
print(url)

# Writing informations in comments:
out_file.write(line_comment + 79*"-" + trail_line_comment + "\n")
out_file.write(line_comment + " " + out_file_name + trail_line_comment + "\n")
FILE_HEADER = line_comment + " Automatically generated from " + str(ARGS.y[0]) \
              + " CODATA NIST file:" + trail_line_comment + "\n" \
              + line_comment + " " + url + trail_line_comment + "\n" \
              + line_comment + " downloaded on " + datetime.date.today().isoformat() + trail_line_comment + "\n"
out_file.write(FILE_HEADER)
out_file.write(line_comment + 79*"-" + trail_line_comment + "\n")

if fortran:
    # Beginning the Fortran module:
    out_file.write("\nmodule " + MODULE_NAME + "\n" + TABS +
                "use, intrinsic :: iso_fortran_env, only: wp=>real64\n" +
                TABS + "implicit none\n\n")

    # Creating a test file that will generate a txt file from the Fortran constants:
    test_file = open(folder_out + "test/test.f90", "w")
    TEST_MODULE_NAME = "test"
    test_file.write("\nmodule " + TEST_MODULE_NAME + "\n" + TABS +
                "use CODATA_constants\n" + TABS +
                "use functions, only: write_CODATA\n" + TABS +
                "use, intrinsic :: iso_fortran_env, only: wp=>real64\n" +
                TABS + "implicit none\n\ncontains\n" +
                "subroutine tests\ncharacter(125) :: s\n" +
                "open(unit = 1, file = 'fortran_generated_"+str(ARGS.y[0])+".txt')\n")

#-------------------------------------------------------------------------------
# Format description of the NIST allascii.txt file:
#
# Header:   11 lines in the 2018 version, 10 lines in the 2010 & 2014 versions,
#           last line being composed of 125 "-"
# Column 1: Quantity = constant name (60 characters)
# Column 2: Value (25 characters)
# Column 3: Uncertainty (25 characters)
# Column 4: Unit (<=15 characters)
#-------------------------------------------------------------------------------

header = True # changes to False later when the header is over
nb_constants = 0

for line in lines_list:
    if ARGS.y[0] == 2010:
        line2 = line.replace("\r\n", "")
    else:
        line2 = line.replace("\n", "")

    if header:
        if fortran:
            # The Fortran test file will print the same header:
            test_file.write("s='"+line2.rstrip()+"'\n")
            test_file.write("write(1,'(a)') trim(s)\n")
        # Looking for the end of the header:
        if line2 == 125*'-':
            header = False
    else: # Header is over
        # Extracting the four columns (with Python the right bound is excluded):
        quantity = line2[0:60].rstrip()
        value = line2[60:85].rstrip()
        uncertainty = line2[85:110].rstrip()
        unit = line2[110:]

        # A special 2014 case:  "{220} lattice spacing of silicon"
        # => a variable name must begin by a letter
        quantity = quantity.replace("{220}", "twotwozero")
        # Variable names must contain only letters, underscores and numerals.
        # => all other characters will be replaced by underscores.
        quantity = quantity.replace(" ", "_").replace("-", "_").replace("/", "_")
        quantity = quantity.replace(".", "_").replace(",", "_").replace("(", "")
        quantity = quantity.replace(")", "").replace("{", "").replace("}", "")
        value = value.replace(" ", "")

        # Deleting the calculated constants?
        if value.find("...") != -1:
            if ARGS.d:
                continue
            else:
                value = value.replace("...", "")

        if fortran:
            # Some constants are mathematically integers, others have a decimal point:
            if value.find(".") == -1 and value.find("e") == -1:
                value = value + "e0"

            # Test: print the value in a txt file, trying to mimick the CODATA format:
            result = get_number_parts(value)
            #print(value, result)
            left = result[0]
            right = result[1]
            expo = result[2]
            test_file.write("call write_CODATA('"+line2[0:60]+"', &\n" + 
                            TABS +"& " +quantity+", "+str(left)+
                            ", "+str(right)+", "+str(expo)+")\n")

        # Declaration in the CODATA_constants file:
        out_file.write(const_prefix + quantity + " = " + value + const_suffix + line_comment + " " + uncertainty + "  " + unit + trail_line_comment + "\n")

        nb_constants = nb_constants + 1

# Finalizing the output file:
out_file.write(final_line)
out_file.close()

print(str(nb_constants)+" constants written")
print('{}{}'.format(os.stat(folder_out+out_file_name).st_size, " bytes"))

if fortran:
    # Finalizing the test file:
    test_file.write("close(1)\n" +
                    "end subroutine tests\nend module "+TEST_MODULE_NAME+"\n")
    test_file.close()
